{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up data directory, data set names, and external data paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "dataDirectory = os.path.abspath(os.path.join(\"..\",\"..\",\"data\", \"okita\", \"main_seCLIP_analysis\"))\n",
    "bindingProteins = [\"RBP_A\", \"RBP_I\", \"RBP_J\", \"RBP_K\"]\n",
    "repsByProtein = {\"RBP_A\" : [\"rep1\", \"rep2\", \"rep3\"],\n",
    "                 \"RBP_I\" : [\"rep1\", \"rep2\", \"rep3_combined\"],\n",
    "                 \"RBP_J\" : [\"rep1_combined\", \"rep2\", \"rep3_combined\"],\n",
    "                 \"RBP_K\" : [\"rep1\", \"rep2\", \"rep3\"]}\n",
    "\n",
    "annotatedGenesFilePath = os.path.abspath(os.path.join(dataDirectory,\"..\",\"__external_data\",\"all.locus_brief_info.7.0_sorted.tsv\"))\n",
    "exonsFilePath = os.path.abspath(os.path.join(dataDirectory,\"..\",\"__external_data\",\"Osativa7_exons.bed\"))\n",
    "genomeFastaFilePath = os.path.abspath(os.path.join(dataDirectory,\"..\",\"__external_data\",\"Osativa7.fa\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate Peaks\n",
    "Note: Make sure all seCLIP .narrowPeak files have their headers removed and are sorted before running this step!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Working with RBP_A_rep1.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_A_rep2.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_A_rep3.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_I_rep1.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_I_rep2.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_I_rep3_combined.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_J_rep1_combined.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_J_rep2.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_J_rep3_combined.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_K_rep1.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_K_rep2.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n",
      "\n",
      "Working with RBP_K_rep3.narrowPeak\n",
      "Checking for encompassing genes...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Checking for encompassing exons...\n",
      "Checking input files for proper sorting...\n",
      "Checking encompassed features file for proper sorting...\n",
      "Checking encompassing features file for proper sorting...\n",
      "Counting in Chr1\n",
      "Counting in Chr10\n",
      "Counting in Chr11\n",
      "Counting in Chr12\n",
      "Counting in Chr2\n",
      "Counting in Chr3\n",
      "Counting in Chr4\n",
      "Counting in Chr5\n",
      "Counting in Chr6\n",
      "Counting in Chr7\n",
      "Counting in Chr8\n",
      "Counting in Chr9\n",
      "Counting in ChrSy\n",
      "Counting in ChrUn\n",
      "Sorting output...\n",
      "Adding peak sequences...\n"
     ]
    }
   ],
   "source": [
    "from AnnotatePeaks import annotatePeaks\n",
    "for bindingProtein in bindingProteins:\n",
    "    for rep in repsByProtein[bindingProtein]:\n",
    "        annotatePeaks([os.path.join(dataDirectory,bindingProtein,f\"{bindingProtein}_{rep}.narrowPeak\")],\n",
    "                      annotatedGenesFilePath, exonsFilePath, genomeFastaFilePath, True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find common loci between repetitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding unique loci in RBP_A_rep1_full_annotation.bed...\n",
      "Finding unique loci in RBP_A_rep2_full_annotation.bed...\n",
      "Finding unique loci in RBP_A_rep3_full_annotation.bed...\n",
      "Writing results...\n",
      "Finding unique loci in RBP_I_rep1_full_annotation.bed...\n",
      "Finding unique loci in RBP_I_rep2_full_annotation.bed...\n",
      "Finding unique loci in RBP_I_rep3_combined_full_annotation.bed...\n",
      "Writing results...\n",
      "Finding unique loci in RBP_J_rep1_combined_full_annotation.bed...\n",
      "Finding unique loci in RBP_J_rep2_full_annotation.bed...\n",
      "Finding unique loci in RBP_J_rep3_combined_full_annotation.bed...\n",
      "Writing results...\n",
      "Finding unique loci in RBP_K_rep1_full_annotation.bed...\n",
      "Finding unique loci in RBP_K_rep2_full_annotation.bed...\n",
      "Finding unique loci in RBP_K_rep3_full_annotation.bed...\n",
      "Writing results...\n"
     ]
    }
   ],
   "source": [
    "from FindCommonLoci import findCommonLoci\n",
    "for bindingProtein in bindingProteins:\n",
    "    fullAnnotationFilePaths = list()\n",
    "    for rep in repsByProtein[bindingProtein]:\n",
    "        fullAnnotationFilePaths.append(\n",
    "            os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_{rep}_full_annotation.bed\")\n",
    "        )\n",
    "    findCommonLoci(fullAnnotationFilePaths, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_common_loci.tsv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (Maybe pool common loci here later...)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use the common loci file to format read sequences for STREME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering requested. Finding valid loci...\n",
      "Found 185 valid loci.\n",
      "Writing positions from RBP_A_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 787 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 185 valid loci.\n",
      "Writing positions from RBP_A_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 587 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 185 valid loci.\n",
      "Writing positions from RBP_A_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_A_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 782 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 241 valid loci.\n",
      "Writing positions from RBP_I_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 992 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 241 valid loci.\n",
      "Writing positions from RBP_I_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 493 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 241 valid loci.\n",
      "Writing positions from RBP_I_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_I_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 990 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 273 valid loci.\n",
      "Writing positions from RBP_J_rep1_combined_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 825 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 273 valid loci.\n",
      "Writing positions from RBP_J_rep1_combined_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 380 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 273 valid loci.\n",
      "Writing positions from RBP_J_rep1_combined_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_J_rep3_combined_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 824 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 410 valid loci.\n",
      "Writing positions from RBP_K_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 1950 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 410 valid loci.\n",
      "Writing positions from RBP_K_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 1122 sequences!\n",
      "Filtering requested. Finding valid loci...\n",
      "Found 410 valid loci.\n",
      "Writing positions from RBP_K_rep1_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep2_full_annotation.bed to intermediate bed file...\n",
      "Writing positions from RBP_K_rep3_full_annotation.bed to intermediate bed file...\n",
      "Finished writitng 1942 sequences!\n"
     ]
    }
   ],
   "source": [
    "from FormatReadSequencesForSTREME import formatReadSequencesForSTREME\n",
    "for bindingProtein in bindingProteins:\n",
    "    fullAnnotationFilePaths = list()\n",
    "    for rep in repsByProtein[bindingProtein]:\n",
    "        fullAnnotationFilePaths.append(\n",
    "            os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_{rep}_full_annotation.bed\")\n",
    "        )\n",
    "    # Only filter on common loci in at least 2 files.\n",
    "    formatReadSequencesForSTREME(\n",
    "        fullAnnotationFilePaths, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_STREME_input_minimal_filtering.fa\"),\n",
    "        genomeFastaFilePath, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_common_loci.tsv\"), 2\n",
    "    )\n",
    "    # Filter on common loci in at least 2 files and sequences <= 50 base pairs. Also expand sequences under 20 base pairs.\n",
    "    formatReadSequencesForSTREME(\n",
    "        fullAnnotationFilePaths, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_STREME_input_20-50bp.fa\"),\n",
    "        genomeFastaFilePath, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_common_loci.tsv\"), 2, 50, 20\n",
    "    )\n",
    "    # Filter on common loci in at least 2 files. Also expand sequences under 20 base pairs.\n",
    "    formatReadSequencesForSTREME(\n",
    "        fullAnnotationFilePaths, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_STREME_input_20bp_min.fa\"),\n",
    "        genomeFastaFilePath, os.path.join(dataDirectory, bindingProtein, f\"{bindingProtein}_common_loci.tsv\"), 2,\n",
    "        minSequenceLength=20\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
